File: src/main/java/me/alegian/thavma/impl/init/registries/deferred/Aspects.java
package me.alegian.thavma.impl.init.registries.deferred;

import me.alegian.thavma.impl.Thavma;
import me.alegian.thavma.impl.common.aspect.Aspect;
import me.alegian.thavma.impl.init.registries.deferred.util.DeferredAspect;
import me.alegian.thavma.impl.init.registries.deferred.util.T7DeferredRegister;

import java.util.List;
import java.util.function.Supplier;

public class Aspects {
  public static final T7DeferredRegister.Aspects REGISTRAR = T7DeferredRegister.createAspects(Thavma.MODID);

  // PRIMAL
  public static final DeferredAspect<Aspect> IGNIS = Aspects.register("ignis", 0xffff5a01, null);
  public static final DeferredAspect<Aspect> AER = Aspects.register("aer", 0xffffff7e, null);
  public static final DeferredAspect<Aspect> LUX = Aspects.register("lux", 0xffffffc0, List.of(Aspects.AER, Aspects.IGNIS));
  public static final DeferredAspect<Aspect> TERRA = Aspects.register("terra", 0xff56c000, null);
  public static final DeferredAspect<Aspect> VITREUS = Aspects.register("vitreus", 0xff80ffff, List.of(Aspects.TERRA, Aspects.AER));
  public static final DeferredAspect<Aspect> AQUA = Aspects.register("aqua", 0xff3cd4fc, null);
  public static final DeferredAspect<Aspect> VICTUS = Aspects.register("victus", 0xffde0005, List.of(Aspects.TERRA, Aspects.AQUA));
  public static final DeferredAspect<Aspect> HERBA = Aspects.register("herba", 0xff01ac00, List.of(Aspects.VICTUS, Aspects.TERRA));
  public static final DeferredAspect<Aspect> ORDO = Aspects.register("ordo", 0xffd5d4ec, null);
  public static final DeferredAspect<Aspect> MOTUS = Aspects.register("motus", 0xffcdccf4, List.of(Aspects.AER, Aspects.ORDO));
  public static final DeferredAspect<Aspect> VOLATUS = Aspects.register("volatus", 0xffe7e7d7, List.of(Aspects.AER, Aspects.MOTUS));
  public static final DeferredAspect<Aspect> BESTIA = Aspects.register("bestia", 0xff9f6409, List.of(Aspects.MOTUS, Aspects.VICTUS));
  public static final DeferredAspect<Aspect> METALLUM = Aspects.register("metallum", 0xffb5b5cd, List.of(Aspects.TERRA, Aspects.ORDO));
  public static final DeferredAspect<Aspect> POTENTIA = Aspects.register("potentia", 0xffc0ffff, List.of(Aspects.ORDO, Aspects.IGNIS));
  // TERTIARY
  public static final DeferredAspect<Aspect> PRAECANTATIO = Aspects.register("praecantatio", 0xffcf00ff, List.of(Aspects.POTENTIA, Aspects.AER));
  public static final DeferredAspect<Aspect> AURAM = Aspects.register("auram", 0xffffc0ff, List.of(Aspects.PRAECANTATIO, Aspects.AER));
  public static final DeferredAspect<Aspect> ALKIMIA = Aspects.register("alkimia", 0xff23ac9d, List.of(Aspects.PRAECANTATIO, Aspects.AQUA));
  public static final DeferredAspect<Aspect> INSTRUMENTUM = Aspects.register("instrumentum", 0xff4040ee, List.of(Aspects.METALLUM, Aspects.POTENTIA));
  public static final DeferredAspect<Aspect> MACHINA = Aspects.register("machina", 0xff8080a0, List.of(Aspects.MOTUS, Aspects.INSTRUMENTUM));
  public static final DeferredAspect<Aspect> PERDITIO = Aspects.register("perditio", 0xff404040, null);
  // DO NOT USE THIS TO CHECK IF AN ASPECT IS PRIMAL, this array is used to auto-generate ores and other content. Instead, use Aspect::isPrimal
  public static final List<DeferredAspect<Aspect>> PRIMAL_ASPECTS = List.of(Aspects.IGNIS, Aspects.AER, Aspects.TERRA, Aspects.AQUA, Aspects.ORDO, Aspects.PERDITIO);
  // SECONDARY (PRIMAL + PRIMAL)
  public static final DeferredAspect<Aspect> VACUOS = Aspects.register("vacuos", 0xff888888, List.of(Aspects.AER, Aspects.PERDITIO));
  public static final DeferredAspect<Aspect> TENEBRAE = Aspects.register("tenebrae", 0xff222222, List.of(Aspects.VACUOS, Aspects.LUX));
  public static final DeferredAspect<Aspect> ALIENIS = Aspects.register("alienis", 0xff805080, List.of(Aspects.VACUOS, Aspects.TENEBRAE));
  public static final DeferredAspect<Aspect> GELUM = Aspects.register("gelum", 0xffe1ffff, List.of(Aspects.IGNIS, Aspects.PERDITIO));
  public static final DeferredAspect<Aspect> MORTUUS = Aspects.register("mortuus", 0xff6a0005, List.of(Aspects.AQUA, Aspects.PERDITIO));
  public static final DeferredAspect<Aspect> SPIRITUS = Aspects.register("spiritus", 0xffebebfb, List.of(Aspects.VICTUS, Aspects.MORTUUS));
  public static final DeferredAspect<Aspect> COGNITIO = Aspects.register("cognitio", 0xfff9967f, List.of(Aspects.IGNIS, Aspects.SPIRITUS));
  public static final DeferredAspect<Aspect> SENSUS = Aspects.register("sensus", 0xffc0ffc0, List.of(Aspects.AER, Aspects.SPIRITUS));
  public static final DeferredAspect<Aspect> AVERSIO = Aspects.register("aversio", 0xffc05050, List.of(Aspects.SPIRITUS, Aspects.PERDITIO));
  public static final DeferredAspect<Aspect> PRAEMUNIO = Aspects.register("praemunio", 0xff00c0c0, List.of(Aspects.SPIRITUS, Aspects.TERRA));
  public static final DeferredAspect<Aspect> DESIDERIUM = Aspects.register("desiderium", 0xffe6be44, List.of(Aspects.SPIRITUS, Aspects.VACUOS));
  public static final DeferredAspect<Aspect> HUMANUS = Aspects.register("humanus", 0xffffd7c0, List.of(Aspects.SPIRITUS, Aspects.VICTUS));
  public static final DeferredAspect<Aspect> EXANIMIS = Aspects.register("exanimis", 0xff3a4000, List.of(Aspects.MOTUS, Aspects.MORTUUS));
  public static final DeferredAspect<Aspect> PERMUTATIO = Aspects.register("permutatio", 0xff578357, List.of(Aspects.PERDITIO, Aspects.ORDO));
  public static final DeferredAspect<Aspect> FABRICO = Aspects.register("fabrico", 0xff809d80, List.of(Aspects.PERMUTATIO, Aspects.INSTRUMENTUM));
  public static final DeferredAspect<Aspect> VITIUM = Aspects.register("vitium", 0xff800080, List.of(Aspects.PERDITIO, Aspects.PRAECANTATIO));
  public static final DeferredAspect<Aspect> VINCULUM = Aspects.register("vinculum", 0xff9a8080, List.of(Aspects.MOTUS, Aspects.PERDITIO));

  private static DeferredAspect<Aspect> register(String id, int color, List<Supplier<Aspect>> components) {
    return Aspects.REGISTRAR.registerAspect(id, () -> new Aspect(id, color, components));
  }
}

File: src/main/java/me/alegian/thavma/impl/init/registries/deferred/Aspects.kt
package me.alegian.thavma.impl.init.registries.deferred

import me.alegian.thavma.impl.Thavma
import me.alegian.thavma.impl.common.aspect.Aspect
import me.alegian.thavma.impl.init.registries.deferred.util.DeferredAspect
import me.alegian.thavma.impl.init.registries.deferred.util.T7DeferredRegister
import java.util.function.Supplier

object Aspects {
    val REGISTRAR: T7DeferredRegister.Aspects = T7DeferredRegister.createAspects(Thavma.MODID)

    // PRIMAL
    val AER = register("aer", 0xffffff7e.toInt(), null)
    val TERRA = register("terra", 0xff56c000.toInt(), null)
    val IGNIS = register("ignis", 0xffff5a01.toInt(), null)
    val AQUA = register("aqua", 0xff3cd4fc.toInt(), null)
    val ORDO = register("ordo", 0xffd5d4ec.toInt(), null)
    val PERDITIO = register("perditio", 0xff404040.toInt(), null)

    // DO NOT use this to check if an aspect is primal. This array is used to auto-generate ores and other content. Instead, use Aspect::isPrimal
    val PRIMAL_ASPECTS = listOf(IGNIS, AER, TERRA, AQUA, ORDO, PERDITIO)

    // SECONDARY
    val VACUOS = register("vacuos", 0xff888888.toInt(), listOf(AER, PERDITIO))
    val LUX = register("lux", 0xffffffc0.toInt(), listOf(AER, IGNIS))
    val MOTUS = register("motus", 0xffcdccf4.toInt(), listOf(AER, ORDO))
    val GELUM = register("gelum", 0xffe1ffff.toInt(), listOf(IGNIS, PERDITIO))
    val VITREUS = register("vitreus", 0xff80ffff.toInt(), listOf(TERRA, AER))
    val METALLUM = register("metallum", 0xffb5b5cd.toInt(), listOf(TERRA, ORDO))
    val VICTUS = register("victus", 0xffde0005.toInt(), listOf(TERRA, AQUA))
    val MORTUUS = register("mortuus", 0xff6a0005.toInt(), listOf(AQUA, PERDITIO))
    val POTENTIA = register("potentia", 0xffc0ffff.toInt(), listOf(ORDO, IGNIS))
    val PERMUTATIO = register("permutatio", 0xff578357.toInt(), listOf(PERDITIO, ORDO))

    // TERTIARY
    val PRAECANTATIO = register("praecantatio", 0xffcf00ff.toInt(), listOf(POTENTIA, AER))
    val AURAM = register("auram", 0xffffc0ff.toInt(), listOf(PRAECANTATIO, AER))
    val ALKIMIA = register("alkimia", 0xff23ac9d.toInt(), listOf(PRAECANTATIO, AQUA))
    val VITIUM = register("vitium", 0xff800080.toInt(), listOf(PERDITIO, PRAECANTATIO))
    val TENEBRAE = register("tenebrae", 0xff222222.toInt(), listOf(VACUOS, LUX))
    val ALIENIS = register("alienis", 0xff805080.toInt(), listOf(VACUOS, TENEBRAE))
    val VOLATUS = register("volatus", 0xffe7e7d7.toInt(), listOf(AER, MOTUS))
    val HERBA = register("herba", 0xff01ac00.toInt(), listOf(VICTUS, TERRA))
    val INSTRUMENTUM = register("instrumentum", 0xff4040ee.toInt(), listOf(METALLUM, POTENTIA))
    val FABRICO = register("fabrico", 0xff809d80.toInt(), listOf(PERMUTATIO, INSTRUMENTUM))
    val MACHINA = register("machina", 0xff8080a0.toInt(), listOf(MOTUS, INSTRUMENTUM))
    val VINCULUM = register("vinculum", 0xff9a8080.toInt(), listOf(MOTUS, PERDITIO))
    val SPIRITUS = register("spiritus", 0xffebebfb.toInt(), listOf(VICTUS, MORTUUS))
    val COGNITIO = register("cognitio", 0xfff9967f.toInt(), listOf(IGNIS, SPIRITUS))
    val SENSUS = register("sensus", 0xffc0ffc0.toInt(), listOf(AER, SPIRITUS))
    val AVERSIO = register("aversio", 0xffc05050.toInt(), listOf(SPIRITUS, PERDITIO))
    val PRAEMUNIO = register("praemunio", 0xff00c0c0.toInt(), listOf(SPIRITUS, TERRA))
    val DESIDERIUM = register("desiderium", 0xffe6be44.toInt(), listOf(SPIRITUS, VACUOS))
    val EXANIMIS = register("exanimis", 0xff3a4000.toInt(), listOf(MOTUS, MORTUUS))
    val BESTIA = register("bestia", 0xff9f6409.toInt(), listOf(MOTUS, VICTUS))
    val HUMANUS = register("humanus", 0xffffd7c0.toInt(), listOf(SPIRITUS, VICTUS))

    private fun register(id: String, color: Int, components: List<Supplier<Aspect>>?): DeferredAspect<Aspect> {
        return REGISTRAR.registerAspect(id) { Aspect(id, color, components) }
    }
}

fun <T> linkedMapWithPrimalKeys(mapper: (DeferredAspect<Aspect>) -> T): LinkedHashMap<DeferredAspect<Aspect>, T> {
    return linkedMapOf(*Aspects.PRIMAL_ASPECTS.map { Pair(it, mapper(it)) }.toTypedArray())
}

fun <T> listFromPrimals(mapper: (DeferredAspect<Aspect>) -> T): List<T> {
    return Aspects.PRIMAL_ASPECTS.map { mapper(it) }
}
